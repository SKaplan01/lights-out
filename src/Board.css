.Board {
  margin: 0 auto;
}

.boardContainer {
  margin-left: 50px;
}

.rowContainer {
  display: flex;
}
