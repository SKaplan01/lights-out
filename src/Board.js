import React, { Component } from 'react';
import Cell from './Cell';
import './Board.css';

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - hasWon: boolean, true when board is all off
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

class Board extends Component {
  constructor(props) {
    super(props);

    // TODO: set initial state
    this.state = { hasWon: false, board: this.createBoard() };
  }

  // difficulty --> an integer that represents how many "clicks" will happen to render the board from a solved board
  static defaultProps = { nrows: 5, ncols: 5, difficulty: 4 };

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */

  createBoard() {
    let board = [];
    // TODO: create solved board (array-of-arrays of false values)
    for (let i = 0; i < this.props.ncols; i++) {
      let row = [];
      for (let j = 0; j < this.props.nrows; j++) {
        row.push(false);
      }
      board.push(row);
    }

    // "click" the board n times (where n is difficulty level)
    for (let i = 0; i < parseInt(this.props.difficulty); i++) {
      //generate a random coordinate
      let y = Math.floor(Math.random() * parseInt(this.props.ncols));
      let x = Math.floor(Math.random() * parseInt(this.props.nrows));
      let coord = `${y}-${x}`;
      this.flipCellsAround(coord, board);
    }

    return board;
  }

  /** handle changing a cell: update board & determine if winner */
  flipCell(y, x, board) {
    let { ncols, nrows } = this.props;
    // if this coord is actually on board, flip it

    if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
      board[y][x] = !board[y][x];
    }
  }

  flipCellsAround(coord, board) {
    let [y, x] = coord.split('-').map(Number);

    // TODO: flip this cell and the cells around it
    this.flipCell(y, x, board);
    this.flipCell(y, x + 1, board);
    this.flipCell(y, x - 1, board);
    this.flipCell(y - 1, x, board);
    this.flipCell(y + 1, x, board);
  }
  // win when every cell is turned off
  // TODO: determine is the game has been won

  checkForWin() {
    let { hasWon, board } = this.state;
    hasWon = board.every(row => {
      return row.every(cell => !cell);
    });

    this.setState({ board, hasWon });
  }

  afterClick(coord, board) {
    this.flipCellsAround(coord, board);
    this.checkForWin();
  }

  /** Render game board or winning message. */

  render() {
    let rows = [];
    for (let y = 0; y < this.props.ncols; y++) {
      let cellsArray = [];
      for (let x = 0; x < this.props.nrows; x++) {
        cellsArray.push(
          <Cell
            key={`${y}-${x}`}
            isLit={this.state.board[y][x]}
            flipCellsAround={() =>
              this.afterClick(`${y}-${x}`, this.state.board)
            }
          />
        );
      }
      rows.push(
        <div className="rowContainer" key={y}>
          {cellsArray}
        </div>
      );
    }
    return (
      // if the game is won, just show a winning msg & render nothing else
      // TODO
      <div className="boardContainer">
        {this.state.hasWon ? 'You won' : rows}
      </div>
      // make table board
      // TODO
    );
  }
}

export default Board;
