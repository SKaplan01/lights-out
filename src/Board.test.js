import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Board from './Board';

it('renders without crashing', function() {
  shallow(<Board />);
});

it('matches snapshot', function() {
  let wrapper = shallow(<Board chanceLightStartsOn={0} />);
  let serialized = toJson(wrapper);
  expect(serialized).toMatchSnapshot();
});

it('handles click', function() {
  let wrapper = mount(<Board nrows={3} ncols={3} chanceLightStartsOn={0} />);

  let elem = wrapper.find('.Cell').first();
  elem.simulate('click');
  console.log('wrapper state', wrapper.state());
  expect(wrapper.state().board[0][0]).toBe(true);
  expect(wrapper.state().board[0][1]).toBe(true);
  expect(wrapper.state().board[1][0]).toBe(true);
});
