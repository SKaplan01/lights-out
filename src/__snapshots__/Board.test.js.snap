// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="boardContainer"
>
  <div
    key="0"
  >
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="0-0"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="0-1"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="0-2"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="0-3"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="0-4"
    />
  </div>
  <div
    key="1"
  >
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="1-0"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="1-1"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="1-2"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="1-3"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="1-4"
    />
  </div>
  <div
    key="2"
  >
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="2-0"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="2-1"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="2-2"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="2-3"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="2-4"
    />
  </div>
  <div
    key="3"
  >
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="3-0"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="3-1"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="3-2"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="3-3"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="3-4"
    />
  </div>
  <div
    key="4"
  >
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="4-0"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="4-1"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="4-2"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="4-3"
    />
    <Cell
      flipCellsAround={[Function]}
      isLit={false}
      key="4-4"
    />
  </div>
</div>
`;
